File: biology/biology.py
import random
from biology_constants import *

class Biology:

    def __init__(self, debug=False):
        """
        Initialize a creatures biology with random values
        :param debug: True to have default genetic values
        """
        self._energy = 1
        self._health = 1
        self._water = 1
        self._genetic_values = GENETIC_VALUES
        if not debug:
            self.set_random_attributes()

    def set_random_attributes(self):
        """
        Sets the genetic values to random values between 0 and 1
        """
        for key in list(self._genetic_values.keys()):
            self._genetic_values[key] = random.random() ** 2

    def display_genetic_vals(self):
        """
        Outputs all of the genetic values for the biology
        """
        for k, v in self._genetic_values.items():
            print(k,": ",v)

    def add_health(self, val):
        """
        Adds an amount of health to the biology
        :param val: the amount to add
        """
        self._health = min(self._health + val, 1.0)

    def add_energy(self, val):
        """
        Adds an amount of energy to the biology
        :param val: Float, the amount to add
        """
        self._energy = min(self._energy + val, 1.0)

    def add_water(self, val):
        """
        Adds an amount of water to the biology
        :param val: Float, The amount to add
        """
        self._water = min(self._health + val, 1.0)

    def get_health(self):
        """
        Returns the current health of the organism
        """
        return self._health

    def get_energy(self):
        """
        Returns the current energy of the organism
        :return: Float, The amount of energy the creature currently has
        """
        return self._energy

    def get_water(self):
        """
        Returns the current water level of the organism
        :return: Float, The amount of water the creature currently has
        """
        return self._water

    def get_efficiencies(self):
        """
        Returns all of the creatures genetic values (efficiencies)
        :return: Dictionary, contains all efficiency values
        """
        return self._genetic_values

    def get_efficiency(self, efficiency):
        """
        Gets a particular efficiency value from the creature
        :param efficiency: The value to query
        :return: Float, that parameters value
        """
        try:
            return self._genetic_values[efficiency]
        except:
            raise KeyError

    def eat_energy(self, quantity):
        """
        Adjusts the energy stores obtained from eating based on the creatures
        various efficiencies.
        :param quantity: The original amount of energy
        :return: Float, the net amount
        """
        amount = quantity * self._genetic_values["Energy Efficiency"]
        self.add_energy(amount)
        return amount

    def drink_water(self, quantity):
        """
        Adjusts the water reserves a creature obtains from drinking based on
        efficiency
        :param quantity: The original amount of water
        :return: Float, the net adjusted amount
        """
        amount = quantity * self._genetic_values["Water Efficiency"]
        self.add_water(amount)
        return amount

    def movement_energy_drain(self, terrain_type):
        """
        Drains energy based on the type of terrain the creature moved through
        :param terrain_type: The particular efficiency to use
        :return: The amount of energy that movement drained
        """
        try:
            efficiency = 1-self._genetic_values[terrain_type]
        except KeyError:
            efficiency = 1-self._genetic_values['Traversal Efficiency 1']
        amount = max(efficiency * TERRAIN_ENERGY_COEFFICIENT
                     * self._genetic_values['Mass'], .01)
        self.add_energy(amount * -1)
        return amount

    def movement_water_drain(self, terrain_type):
        """
        Drains water based on the type of terrain navifated
        :param terrain_type: The particular efficiency to use
        :return: the net water drainage
        """
        try:
            efficiency = 1-self._genetic_values[terrain_type]
        except KeyError:
            efficiency = 1-self._genetic_values['Traversal Efficiency 1']
        water_efficiency = 1-self._genetic_values["Water Efficiency"]
        amount = max(water_efficiency * efficiency *
                      TERRAIN_WATER_COEFFICIENT, .01)
        self.add_water(amount * -1)
        return amount

    def add_chemical(self, chemical_type, quantity):
        """
        Adjusts health based on the chemical type passed
        :param chemical_type: Which chemical the creature is absorbing
        :param quantity: The amount the creature is absorbing
        :return: The amount that health was edjusted
        """
        try:
            efficiency = self._genetic_values[chemical_type] - .5
        except:
            raise KeyError
        amount = efficiency * quantity * HEALTH_COEFFICIENT
        self.add_health(amount)
        return amount

    def tick_energy_drain(self):
        """
        Determines how much energy the creature loses every tick
        :return: The net amount drained
        """
        total = 0
        for k, v in self._genetic_values.items():
            if k == 'Mass':
                total += v ** .2
            else:
             total += v
        total = total ** .5 / len(list(self._genetic_values.keys()))
        total = max(total * (1-(self._genetic_values["Mass"]**2)),.02)
        self.add_energy(total * -1 * ENERGY_DRAIN_COEFFICIENT)
        return total

    def tick_health_drain(self):
        """
        Determines how much health the creature loses each tick
        :return: The amount of health drained
        """
        if self._energy < 1 - self._genetic_values["Mass"]:
            difference = self._genetic_values["Mass"] - self._energy
            drain = difference ** 2
            self.add_health(drain * -1)
            return drain
        return 0

    def update(self):
        self._energy = max(self._energy, 0)
        self._water = max(self._water, 0)
        print('Tick energy loss: ', self.tick_energy_drain())
        print('Tick Health loss', self.tick_health_drain())

    def check_death(self):
        if self._health <= 0.0:
            return True
        return False

File: decision_center/biology.cpp
#include "biology.hpp"
#include <iostream>
#include <iomanip>
#include <random>
#include <cmath>
#include <algorithm>

// Random number generator (thread-safe singleton)
static std::random_device rd;
static std::mt19937 gen(rd());
static std::uniform_real_distribution<> dis(0.0, 1.0);

Biology::Biology(bool debug)
    : _energy(1.0), _health(1.0), _water(1.0)
{
    _genetic_values = GetDefaultGeneticValues();
    if (!debug)
    {
        set_random_attributes();
    }
}

// ==================== Initialization & Setup ====================

void Biology::set_random_attributes()
{
    /**
     * Sets the genetic values to random values between 0 and 1
     * Uses the formula: random() ** 2 to bias towards lower values
     */
    for (auto& pair : _genetic_values)
    {
        double random_val = dis(gen);
        pair.second = random_val * random_val;  // Equivalent to random() ** 2
    }
}

// ==================== Getters ====================

double Biology::get_health() const
{
    return _health;
}

double Biology::get_energy() const
{
    return _energy;
}

double Biology::get_water() const
{
    return _water;
}

std::unordered_map<std::string, double> Biology::get_efficiencies() const
{
    return _genetic_values;
}

double Biology::get_efficiency(const std::string& efficiency) const
{
    auto it = _genetic_values.find(efficiency);
    if (it == _genetic_values.end())
    {
        throw std::out_of_range("Genetic value not found: " + efficiency);
    }
    return it->second;
}

std::unordered_map<std::string, double> Biology::get_genetic_vals() const
{
    return _genetic_values;
}

// ==================== Setters ====================

void Biology::set_efficiencies(const std::unordered_map<std::string, double>& vals)
{
    _genetic_values = vals;
}

void Biology::set_efficiency(const std::string& type, double value)
{
    if (value < 0.0 || value > 1.0)
    {
        throw std::invalid_argument("Efficiency value must be between 0 and 1");
    }

    auto it = _genetic_values.find(type);
    if (it == _genetic_values.end())
    {
        throw std::out_of_range("Genetic trait not found: " + type);
    }

    it->second = value;
}

void Biology::add_health(double val)
{
    _health = std::min(_health + val, 1.0);
}

void Biology::add_energy(double val)
{
    _energy = std::min(_energy + val, 1.0);
}

void Biology::add_water(double val)
{
    // NOTE: Original Python had a bug here using _health instead of _water
    // This corrected version uses _water
    _water = std::min(_water + val, 1.0);
}

// ==================== Resource Consumption ====================

double Biology::eat_energy(double quantity)
{
    /**
     * Adjusts the energy stores obtained from eating based on the creature's
     * various efficiencies.
     */
    double amount = quantity * _genetic_values["Energy Efficiency"];
    amount *= std::pow(1.0 - _genetic_values["Mass"], 0.5);
    add_energy(amount);
    return amount;
}

double Biology::drink_water(double quantity)
{
    /**
     * Adjusts the water reserves a creature obtains from drinking based on
     * efficiency.
     */
    double amount = quantity * _genetic_values["Water Efficiency"];
    add_water(amount);
    return amount;
}

double Biology::add_chemical(const std::string& chemical_type, double quantity)
{
    /**
     * Adjusts health based on the chemical type passed.
     */
    auto it = _genetic_values.find(chemical_type);
    if (it == _genetic_values.end())
    {
        throw std::out_of_range("Chemical type not found: " + chemical_type);
    }

    double efficiency = it->second - 0.5;
    double amount = efficiency * quantity * HEALTH_COEFFICIENT;
    add_health(amount);
    return amount;
}

// ==================== Terrain & Movement ====================

double Biology::movement_energy_drain(const std::string& terrain_type)
{
    /**
     * Drains energy based on the type of terrain the creature moved through.
     * Falls back to default traversal efficiency if terrain type not found.
     */
    double efficiency;
    auto it = _genetic_values.find(terrain_type);

    if (it == _genetic_values.end())
    {
        // Use default traversal efficiency
        efficiency = 1.0 - _genetic_values.at("Traversal Efficiency 1");
    }
    else
    {
        efficiency = 1.0 - it->second;
    }

    double amount = std::max(
        efficiency * TERRAIN_ENERGY_COEFFICIENT * _genetic_values["Mass"],
        0.01
    );

    add_energy(amount * -1);
    return amount;
}

double Biology::movement_water_drain(const std::string& terrain_type)
{
    /**
     * Drains water based on the type of terrain navigated.
     */
    double efficiency;
    auto it = _genetic_values.find(terrain_type);

    if (it == _genetic_values.end())
    {
        // Use default traversal efficiency
        efficiency = 1.0 - _genetic_values.at("Traversal Efficiency 1");
    }
    else
    {
        efficiency = 1.0 - it->second;
    }

    double water_efficiency = 1.0 - _genetic_values["Water Efficiency"];
    double amount = std::max(
        water_efficiency * efficiency * TERRAIN_WATER_COEFFICIENT,
        0.01
    );

    add_water(amount * -1);
    return amount;
}

// ==================== Life Cycle ====================

double Biology::tick_energy_drain()
{
    /**
     * Determines how much energy the creature loses every tick.
     * Based on the sum of genetic traits (excluding Mass).
     */
    double total = 0.0;

    for (const auto& pair : _genetic_values)
    {
        if (pair.first != "Mass")
        {
            total += pair.second;
        }
    }

    // Calculate the drain: sqrt of sum, divided by number of traits, adjusted for mass
    total = std::pow(total, 0.5) / static_cast<double>(_genetic_values.size());
    total = std::max(
        total * (1.0 - std::pow(_genetic_values["Mass"], 2.0)),
        0.02
    );

    double drain = total * ENERGY_DRAIN_COEFFICIENT;
    add_energy(drain * -1);
    return total;
}

double Biology::tick_health_drain()
{
    /**
     * Determines how much health the creature loses each tick.
     * Health is drained if energy falls below a mass-dependent threshold.
     */
    if (_energy < 1.0 - _genetic_values["Mass"])
    {
        double difference = _genetic_values["Mass"] - _energy;
        double drain = std::pow(difference, 2.0);
        add_health(drain * -1);
        return drain;
    }
    return 0.0;
}

void Biology::update()
{
    /**
     * Updates the organism for a single game tick.
     * Clamps resources and applies per-tick drains.
     */
    _energy = std::max(_energy, 0.0);
    _water = std::max(_water, 0.0);

    std::cout << "Tick energy loss: " << tick_energy_drain() << std::endl;
    std::cout << "Tick Health loss: " << tick_health_drain() << std::endl;
}

bool Biology::check_death() const
{
    /**
     * Checks if the organism should be considered dead.
     */
    return _health <= 0.0;
}

// ==================== Display & Debugging ====================

void Biology::display_genetic_vals() const
{
    /**
     * Outputs all of the genetic values for the biology.
     */
    for (const auto& pair : _genetic_values)
    {
        std::cout << std::setw(25) << std::left << (pair.first + ":") 
                  << pair.second << std::endl;
    }
}

void Biology::print_vals() const
{
    /**
     * Displays current resource levels.
     */
    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Current Health: " << _health << std::endl;
    std::cout << "Current Energy: " << _energy << std::endl;
    std::cout << "Current Water: " << _water << std::endl;
}
